### PR TITLE
PojoCodec: Fix mapping for nested generic types

### PR DESCRIPTION
**Pull request by rozza**
JAVA-3408

https://evergreen.mongodb.com/version/5e2b10122a60ed70c4811d16

### CODE REVIEW

File: bson/src/main/org/bson/codecs/pojo/PojoSpecializationHelper.java [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+    private static TypeData<?> getTypeData(final TypeData<?> parentTypeData, final List<TypeData<?>> specializedTypeParameters,
+                                           final Map<Integer, Either<Integer, TypeParameterMap>> propertyToClassParamIndexMap,
+                                           final int index) {
+        if (!propertyToClassParamIndexMap.containsKey(index)) {
+            return parentTypeData;
+        }
+        return propertyToClassParamIndexMap.get(index).map(l -> {
+                    if (parentTypeData.getTypeParameters().isEmpty()) {
+                        // Represents the whole typeData
+                        return specializedTypeParameters.get(l);
```
Comment (rozza) [inline bson/src/main/org/bson/codecs/pojo/PojoSpecializationHelper.java]: The `typeData` doesn't have any type parameters, so the specialization must represent the whole type.

File: bson/src/main/org/bson/codecs/pojo/PojoSpecializationHelper.java [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+                    } else {
+                        // Represents a single nested type parameter within this typeData
```
Comment (rozza) [inline bson/src/main/org/bson/codecs/pojo/PojoSpecializationHelper.java]: Its one of the type parameters that needs to be specialized.

File: bson/src/main/org/bson/codecs/pojo/PojoSpecializationHelper.java [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+                        TypeData.Builder<?> builder = TypeData.builder(parentTypeData.getType());
+                        List<TypeData<?>> typeParameters = new ArrayList<>(parentTypeData.getTypeParameters());
+                        typeParameters.set(index, specializedTypeParameters.get(l));
+                        builder.addTypeParameters(typeParameters);
+                        return builder.build();
+                    }
+                },
+                r -> {
+                    // Represents a child type parameter of this typeData
+                    return getTypeData(parentTypeData, specializedTypeParameters, r.getPropertyToClassParamIndexMap());
```
Comment (rozza) [inline bson/src/main/org/bson/codecs/pojo/PojoSpecializationHelper.java]: The `parentTypeData` doesn't need specializing but one of its nested type parameters does, so we recurse down a level.

File: bson/src/main/org/bson/codecs/pojo/PojoSpecializationHelper.java [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+    @SuppressWarnings("unchecked")
+    static <V> TypeData<V> specializeTypeData(final TypeData<V> typeData, final List<TypeData<?>> typeParameters,
+                                              final TypeParameterMap typeParameterMap) {
+        if (!typeParameterMap.hasTypeParameters() || typeParameters.isEmpty()) {
+            return typeData;
+        }
+
+        Map<Integer, Either<Integer, TypeParameterMap>> propertyToClassParamIndexMap = typeParameterMap.getPropertyToClassParamIndexMap();
+        Either<Integer, TypeParameterMap> classTypeParamRepresentsWholeField = propertyToClassParamIndexMap.get(-1);
+        if (classTypeParamRepresentsWholeField != null) {
```
Comment (rozza) [inline bson/src/main/org/bson/codecs/pojo/PojoSpecializationHelper.java]: Handles the case where a generic type represents the whole type.

File: bson/src/main/org/bson/codecs/pojo/PojoSpecializationHelper.java [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+            Integer index = classTypeParamRepresentsWholeField.map(i -> i, e -> {
+                throw new IllegalStateException("Invalid state, the whole class cannot be represented by a subtype.");
+            });
+            return (TypeData<V>) typeParameters.get(index);
+        } else {
+            return getTypeData(typeData, typeParameters, propertyToClassParamIndexMap);
```
Comment (rozza) [inline bson/src/main/org/bson/codecs/pojo/PojoSpecializationHelper.java]: One of the type parameters needs specializing

File: bson/src/main/org/bson/codecs/pojo/PojoSpecializationHelper.java [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+                        TypeData.Builder<?> builder = TypeData.builder(typeData.getType());
+                        List<TypeData<?>> typeParameters = new ArrayList<>(typeData.getTypeParameters());
+                        typeParameters.set(index, specializedTypeParameters.get(l));
+                        builder.addTypeParameters(typeParameters);
+                        return builder.build();
+                    }
+                },
+                r -> {
+                    // Represents a child type parameter of this typeData
+                    return getTypeData(typeData, specializedTypeParameters, r.getPropertyToClassParamIndexMap());
```
Comment (rozza) [inline bson/src/main/org/bson/codecs/pojo/PojoSpecializationHelper.java]: Nothing needs specializing for this type, but one of this types type parameters needs specializing, so we recurse down and specialize there.

File: bson/src/main/org/bson/codecs/pojo/TypeParameterMap.java [hunk 7]
```diff
@@ -26,7 +26,7 @@
  * Maps the index of a class's generic parameter type index to a property's.
  */
 final class TypeParameterMap {
-    private final Map<Integer, Integer> propertyToClassParamIndexMap;
+    private final Map<Integer, Either<Integer, TypeParameterMap>> propertyToClassParamIndexMap;
```
Comment (rozza) [inline bson/src/main/org/bson/codecs/pojo/TypeParameterMap.java]: The `TypeParameterMap` has changed from a flat level map to a nested level map.  This allows the nested generic 
types to be handled (eg: `List<List<T>>`) as we can map the levels effectively.

Essentially it allows  the mapping of:
```
List<T> => {0, 0}
List<List<T>> => {0, {0, 0}}
```

File: bson/src/test/unit/org/bson/codecs/pojo/ClassModelTest.java [hunk 8]
```diff
@@ -55,91 +62,82 @@ public void testSimpleGenericsModel() {
         assertEquals("_t", classModel.getDiscriminatorKey());
         assertEquals("org.bson.codecs.pojo.entities.SimpleGenericsModel", classModel.getDiscriminator());
         assertNull(classModel.getIdPropertyModel());
-        assertEquals(4, classModel.getPropertyModels().size());
         assertTrue(classModel.getInstanceCreatorFactory() instanceof InstanceCreatorFactoryImpl);
+
+        assertEquals(4, classModel.getPropertyModels().size());
```
Comment (rozza) [inline bson/src/test/unit/org/bson/codecs/pojo/ClassModelTest.java]: Lots of changes to improve the testing of all the type data for the property models.

File: bson/src/test/unit/org/bson/codecs/pojo/ClassModelTest.java [hunk 9]
```diff
@@ -178,11 +182,93 @@ public void testInheritedClassAnnotations() {
...
+        ClassModel<?> classModel = ClassModel.builder(ListGenericExtendedModel.class).build();
 
-        ClassModel<?> overriddenImplementationClassModel = ClassModel.builder(ShapeHolderCircleModel.class).build();
         assertEquals(1, classModel.getPropertyModels().size());
-        assertEquals(ShapeModelCircle.class, overriddenImplementationClassModel.getPropertyModels().get(0).getTypeData().getType());
+        assertEquals(createTypeData(List.class, Integer.class), classModel.getPropertyModel("values").getTypeData());
+    }
+
+    @Test
+    public void testListListGenericExtendedModel() {
```
Comment (rozza) [inline bson/src/test/unit/org/bson/codecs/pojo/ClassModelTest.java]: New classes to test

File: bson/src/test/unit/org/bson/codecs/pojo/PojoRoundTripTest.java [hunk 10]
```diff
@@ -445,6 +453,61 @@ public void test() {
                 getPojoCodecProviderBuilder(BsonIgnoreDuplicatePropertyMultipleTypes.class).conventions(Conventions.DEFAULT_CONVENTIONS),
                 "{stringField: 'string value'}"));
 
+        data.add(new TestData("Can handle concrete generic list types",
```
Comment (rozza) [inline bson/src/test/unit/org/bson/codecs/pojo/PojoRoundTripTest.java]: Improved roundTrip coverage for these new classes.

File: bson/src/main/org/bson/codecs/pojo/PojoSpecializationHelper.java [hunk 11]
```diff
@@ -0,0 +1,82 @@
...
+ * limitations under the License.
+ */
+
+package org.bson.codecs.pojo;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+
+final class PojoSpecializationHelper {
```
[the remark this answers is not in the document]
Comment (evanchooly) [reply bson/src/main/org/bson/codecs/pojo/PojoSpecializationHelper.java]: This looks really promising but it's all package private.  I won't be able to access it from Morphia.  Would it make sense to make this an instance method on `TypeData` itself?

[the remark this answers is not in the document]
Comment (rozza) [reply bson/src/main/org/bson/codecs/pojo/PojoSpecializationHelper.java]: It is a helper, so not sure it needs to be an instance method - especially as TypeData is immutable. 

The idea for this fix is it does the right thing based on the ClassModelBuilder and the associated property models passed into it when building the ClassModel.

I'm not sure it warrants becoming part of the public API, or at least I don't understand the need for it to be. What would Morphia do with it if it was public, that the ClassModelBuilder.build() method already doesn't do?

[the remark this answers is not in the document]
Comment (evanchooly) [reply bson/src/main/org/bson/codecs/pojo/PojoSpecializationHelper.java]: I can't actually use `ClassModel`/`ClassModelBuilder` for a  few reasons.  `ClassModel` doesn't seem to expose the discovered annotations which Morphia needs.  Morphia also needs access to the fields which are not exposed either.  My original PR was much larger and included making more things public/protected so I could access them and adding new methods.  Ultimately, I switched to my own `EntityModel` which is much simpler for my needs and more customizable in terms of codec workflow (lifecycle events, e.g.).

All that to say that having `ClassBuilder` have access to it doesn't really help in my builder code.  This method doesn't necessarily need to be an instance method but it at least needs to be public so I can access it.  My only other option is to copy this source in to Morphia and embed it directly which is distinctly unappealing.

[the remark this answers is not in the document]
Comment (rozza) [reply bson/src/main/org/bson/codecs/pojo/PojoSpecializationHelper.java]: OK - I think that helps me understand Morphia's needs better.  Lets continue the discussion on your PR.  This has now been merged into master.

**Pull request closed without merging**